include/cupla/math/Common.hpp

```diff
@@ -0,0 +1,166 @@
+    -> alpaka::concepts::ImplementationBase<
+        T_Concept,
+        Acc
+    >
+    {
+        using AccThreadSeqMathConcept = alpaka::concepts::ImplementationBase<
+            T_Concept,
+            AccThreadSeq
+        >;
+        using AccMathConcept = alpaka::concepts::ImplementationBase<
+            T_Concept,
+            Acc
+        >;
+        // cupla Acc and AccThreadSeq should use the same math concept implementation
+        static_assert(
+            std::is_same<
+                AccMathConcept,
+                AccThreadSeqMathConcept
+            >::value,
+            "The math concept implementation for the type 'Acc' and 'AccThreadSeq' must be equal"
+        );
+        return AccMathConcept{};
+    }
+} // namespace detail
+
+/* Using the free alpaka functions `alpaka::math::*` will result into `__host__ __device__`
+ * errors, therefore the alpaka math trait must be used.
+ */
+#define CUPLA_UNARY_MATH_FN(functionName, alpakaMathConcept, alpakaMathTrait)  \
+    /**                                                                        \
+     * @tparam T_Type argument type                                            \
+     * @param arg input argument                                               \
+     */                                                                        \
+    ALPAKA_NO_HOST_ACC_WARNING                                                 \
+    template< typename T_Type >                                                \
+    ALPAKA_FN_ACC ALPAKA_FN_INLINE auto functionName(                          \
+        T_Type const & arg                                                     \
+    )                                                                          \
+    /* return type is required because nvcc can not detect the return type     \
+     * for device functions.                                                   \
+     */                                                                        \
+    ->  decltype(                                                              \
+        alpaka::math::traits::alpakaMathTrait<                                 \
+            alpaka::concepts::ImplementationBase<                              \
+                alpakaMathConcept,                                             \
+                Acc                                                            \
+            >,                                                                 \
+            T_Type                                                             \
+        >::functionName(                                                       \
+            detail::getConcept< alpakaMathConcept >(),                         \
+            arg                                                                \
+        )                                                                      \
+    )                                                                          \
```

Comment:
Is this needed because `auto` deduces incorrectly somehow?

Same question for a similar situation later in this file.

Reply:
Here it is not required but later for `ALPAKA_FN_ACC` is is necessary. nvcc can not detect the return type for `__device__` functions